src/task_impl/std/mod.rs

```diff
@@ -475,14 +475,35 @@ impl ThreadUnpark {
     }
 
     fn park(&self) {
-        if !self.ready.swap(false, Ordering::SeqCst) {
-            thread::park();
+        // If `ready` is true, that means we need to immediately re-poll, so we
+        // short-circuit instead of parking.
+        if self.ready.swap(false, Ordering::SeqCst) {
+            return;
         }
+
+        // Block until the next `notify`. Note that if a call to `notify` races
+        // in between our check above and this `park`, that's ok, because
+        // `thread::park` sets an internal flag of its own similar to our
+        // `ready`.
+        thread::park();
+
+        // Now we've awoken from `park`. If it happened the usual way, by
+        // another thread calling `notify` while we slept, then the `ready` flag
+        // has been set back to true. Since the wait loop is going to `poll`
+        // again as soon as we return, and there's no need to `poll` twice, we
+        // clear it.
+        self.ready.store(false, Ordering::SeqCst);
     }
 }
 
 impl Notify for ThreadUnpark {
     fn notify(&self, _unpark_id: usize) {
+        // The target thread might already be in the middle of `poll` when
+        // `notify` is called. In that case it might've just missed the
+        // readiness of some part of its future, and it needs to immediately
+        // call `poll` all over again as soon as the current `poll` is done.
+        // Setting the `ready` flag makes this happen by skipping the next
+        // `park`.
         self.ready.store(true, Ordering::SeqCst);
         self.thread.unpark()
```

Comment:
Thinking out loud about a side issue that probably doesn't matter: If `notify` is called while the target thread is running, it will set _two_ ready flags, both `self.ready` right here and also the buffered unpark flag set by `std::thread::Thread::unpark`. As we've been discussing re: #132, it's possible that that buffered unpark might get eaten by user code at any point. But if that doesn't happen, then I think the result is that `poll` will get called again _twice_:

1. `notify` comes in while a `poll` is running.
2. The wait loop finishes the current `poll` and calls `ThreadUnpark::park`.
3. The `ready` flag is true, so `park` swaps it and short-circuits.
4. The wait loop calls `poll` again and then `ThreadUnpark::park` again. (This satisfies `notify`'s contract.)
5. The `ready` flag is false, so `park` calls `thread::park`. But the "buffered unpark" created in step 1 makes it return immediately (maybe).
6. The wait loop calls `poll` again and then `ThreadUnpark::park` again. (This is arguably wasted work.)
7. Finally `park` calls `thread::park` and blocks.

I don't think this is a big deal. `std::thread::park` is allowed to spuriously return whenever it feels like it anyway. I suppose if we _really_ cared about avoiding this, we could do something like this:

```rust
fn park(&self) {
    // Clear any buffered unparks on the current thread, with a park that's
    // guaranteed to return immediately.
    self.thread.unpark();
    thread::park();

    // Do all the other stuff we were doing before.
    if self.ready.swap(false, Ordering::SeqCst) {
        return;
    }
    thread::park();
    self.ready.store(false, Ordering::SeqCst);
}
```

But I'm not sure that's actually correct, and I don't know how expensive it is, and I don't know if the extra complexity is worth it anyway.

Reply:
(Apologies for all the longwindness here.) Thinking more about reordering atomic operations in the _current_ (prior to this PR) version of this code. Is the compiler/CPU allowed to reorder the `self.ready.store` and the `self.thread.unpark` in our implementation of `ThreadUnpark::notify`? If so, I worry the following race is possible:

1. The main thread is in the middle of a `poll`.
2. The other thread calls `ThreadUnpark::notify`.
3. `notify` calls `self.thread.unpark()` _first_ (assuming this reordering is possible).
4. The main thread does some more work, which (a la #132) happens to consume the buffered unpark.
5. The main thread finishes `poll` and calls `ThreadUnpark::park`.
6. The main thread loads `self.ready` and finds that it's false.
7. Finally control resumes in `notify` on the other thread and (again assuming reordering is possible) sets `self.ready`, too late for the load above to see it.
8. The main thread calls `thread::park()` and blocks because of step 4, violating `notify`'s contract.

Is there some larger rule like "no memory operations will ever get reordered around acquiring or releasing any mutex" that protects us here? I'm afraid I don't know enough about how all these primitives work.

Reply:
An excellent point about the spurious wakeup! It's true that up to this point it hasn't been considered too too serious because most implementations of `poll` are "fast enough" and otherwise spurious wakeups are expected every now and then.

In general though yeah the usage of `SeqCst` should mean that nothing is reordered here too much, in theory.

Reply:
@oconnor663 A mutex is the strongest form of memory ordering, so there won't be any reordering around it.